services/spar/src/Spar/Intra/Brig.hs

```diff
@@ -262,14 +262,31 @@ assertIsTeamOwner buid tid = do
 --
 -- Called by post handler, and by 'authorizeIdP'.
 getZUsrOwnedTeam :: (HasCallStack, SAML.SP m, MonadSparToBrig m)
-            => Maybe UserId -> m TeamId
+                 => Maybe UserId -> m TeamId
 getZUsrOwnedTeam Nothing = throwSpar SparMissingZUsr
 getZUsrOwnedTeam (Just uid) = do
   usr <- getUser uid
   case userTeam =<< usr of
     Nothing -> throwSpar SparNotInTeam
     Just teamid -> teamid <$ assertIsTeamOwner uid teamid
 
+-- | Verify user's password (needed for certain powerful operations).
+ensureReAuthorised :: (HasCallStack, MonadSparToBrig m)
+                   => Maybe UserId -> Maybe PlainTextPassword -> m ()
+ensureReAuthorised Nothing _ = throwSpar SparMissingZUsr
+ensureReAuthorised (Just uid) secret = do
+  resp <- call
+    $ method GET
+    . paths ["/i/users", toByteString' uid, "reauthenticate"]
+    . json (ReAuthUser secret)
+  if | statusCode resp == 200
+       -> pure ()
+     | statusCode resp == 403
+       -> throwSpar SparReAuthRequired
+     | inRange (400, 499) (statusCode resp)
+       -> throwSpar . SparBrigErrorWith (responseStatus resp) $ "reauthentication failed"
```

Comment:
[ignore if you think it's not important] I suspect 400 should be an internal error as well.

Reply:
> an internal error

Do you mean 5xx, or..?

Reply:
yes.

Reply:
If the password can not be decoded as JSON (i.e. if it's shorter than 6 symbols), you'll get a 400 out of the Brig reauthentication endpoint.

We are lucky that in this case we _also_ use the `PlainTextPassword` type and so the undecodable password will be caught before Brig sees it, but if it ever changes for some reason, we'll have a potential 5xx that is actually entirely benign.